### PR TITLE
Cart

### PR DESCRIPTION
**Pull request by ShaunDaneJames**
Shopping Cart and Checkout
- Creates restrictions around a cart. 
- Allows carts to decrease quantity
- Allows cart to decrease quantity
- Registered users can check out.

### CODE REVIEW

**Review by hopesgit — APPROVED** [review]
Nice job dealing with all those newly-failing tests thanks to the orders user_id migration! Wonderful job! Let's get that money!

File: spec/features/cart/show_spec.rb [hunk 1]
```diff
@@ -6,6 +6,12 @@
       before(:each) do
         @mike = Merchant.create(name: "Mike's Print Shop", address: '123 Paper Rd.', city: 'Denver', state: 'CO', zip: 80203)
         @meg = Merchant.create(name: "Meg's Bike Shop", address: '123 Bike Rd.', city: 'Denver', state: 'CO', zip: 80203)
+        @hope = Merchant.create(name: "Hope's Metroid Shop", address: '125 XR42 St.', city: 'Denver', state: 'CO', zip: 80210)
```
Comment (GeorgieGirl24) [inline spec/features/cart/show_spec.rb:9]: Great shop name!

File: spec/features/cart/show_spec.rb [hunk 2]
```diff
@@ -69,7 +75,38 @@
...
+      visit "/items/#{@tire.id}"
+      click_on "Add To Cart"
+      visit "/items/#{@pencil.id}"
+      click_on "Add To Cart"
+      @items_in_cart = [@paper,@tire,@pencil]
+
+      visit '/cart'
+      save_and_open_page
+      within("#cart-item-#{@pencil.id}") do
+        click_link("➕")
```
Comment (GeorgieGirl24) [inline spec/features/cart/show_spec.rb:100]: Great idea for  a symbol and a link!

File: app/models/order.rb [hunk 3]
```diff
@@ -3,7 +3,8 @@ class Order <ApplicationRecord
 
   has_many :item_orders
   has_many :items, through: :item_orders
-
+  enum status: %w(pending packaged shipped cancelled)
```
Comment (GeorgieGirl24) [inline app/models/order.rb:7]: Good insight!

File: app/controllers/cart_controller.rb [hunk 4]
```diff
@@ -21,6 +21,24 @@ def remove_item
     redirect_to '/cart'
   end
 
+  def increment_quantity
+    if cart.contents[params[:item_id]] < Item.find(params[:item_id]).inventory
```
Comment (GeorgieGirl24) [inline app/controllers/cart_controller.rb:25]: Sweet logic here.

File: app/controllers/cart_controller.rb [hunk 5]
```diff
@@ -21,6 +21,24 @@ def remove_item
...
+    else
+      flash[:error] = "Not enough in stock"
+      redirect_to '/cart'
+    end
+  end
+
+  def decrement_quantity
+    cart.contents[params[:item_id]] -= 1
+    if cart.contents[params[:item_id]] == 0
+      session[:cart].delete(params[:item_id])
```
Comment (GeorgieGirl24) [inline app/controllers/cart_controller.rb:37]: Great, so if there is no stock that item is deleted

File: db/migrate/20201031201347_add_status_to_orders.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddStatusToOrders < ActiveRecord::Migration[5.2]
```
Comment (GeorgieGirl24) [inline db/migrate/20201031201347_add_status_to_orders.rb:1]: Hooray! Nice migration. I was looking through the stories and realized that we'd need to add this column. Great job team!

File: db/migrate/20201031221038_add_column_user_id_to_orders.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class AddColumnUserIdToOrders < ActiveRecord::Migration[5.2]
+  def change
```
Comment (GeorgieGirl24) [inline db/migrate/20201031221038_add_column_user_id_to_orders.rb:2]: 👍

File: app/views/orders/show.html.erb [hunk 8]
```diff
@@ -7,13 +7,15 @@
...
       <th>Zip</th>
+      <th>Status</th>
     </tr>
     <tr>
       <td><p><%= @order.name %> </p></td>
       <td><p><%= @order.address %> </p></td>
       <td><p><%= @order.city %> </p></td>
       <td><p><%= @order.state %> </p></td>
       <td><p><%= @order.zip %> </p></td>
+      <td><p><%= @order.status %> </p></td>
```
Comment (GeorgieGirl24) [inline app/views/orders/show.html.erb:18]: great addition

File: app/views/cart/show.html.erb [hunk 9]
```diff
@@ -42,9 +44,15 @@
 </table>
   <p>Total: <%= number_to_currency(cart.total) %></p>
   <p><%= link_to "Empty Cart", "/cart", method: :delete %></p>
-  <p><%= link_to "Checkout", "/orders/new", method: :get %></p>
+  <p><%= link_to "Checkout", "/orders/new", method: :get if current_user %></p>
```
Comment (GeorgieGirl24) [inline app/views/cart/show.html.erb:47]: Nice clause!

File: app/models/user.rb [hunk 10]
```diff
@@ -1,4 +1,5 @@
 class User < ApplicationRecord
+  has_many :orders
```
Comment (GeorgieGirl24) [inline app/models/user.rb:2]: Sweet! I love this new association!

File: spec/models/order_spec.rb [hunk 11]
```diff
@@ -12,17 +12,26 @@
   describe "relationships" do
     it {should have_many :item_orders}
     it {should have_many(:items).through(:item_orders)}
+    it {should belong_to :user}
```
Comment (GeorgieGirl24) [inline spec/models/order_spec.rb:15]: Nice additional relationship established!

File: spec/models/user_spec.rb [hunk 12]
```diff
@@ -10,6 +10,7 @@
     it {should validate_presence_of(:email_address)}
     it {should validate_uniqueness_of(:email_address)}
     it {should validate_presence_of(:password)}
+    it {should have_many :orders}
```
Comment (GeorgieGirl24) [inline spec/models/user_spec.rb:13]: Sweet!

**Review by GeorgieGirl24 — APPROVED** [review]

**Review by GeorgieGirl24 — APPROVED** [review]

**Pull request merged**